zkevm-circuits/src/evm_circuit/util/math_gadget/abs_word.rs

```diff
@@ -0,0 +1,160 @@
+        );
+
+        Self {
+            x,
+            x_abs,
+            sum,
+            is_neg,
+            add_words,
+        }
+    }
+
+    pub(crate) fn assign(
+        &self,
+        region: &mut CachedRegion<'_, '_, F>,
+        offset: usize,
+        x: Word,
+        x_abs: Word,
+    ) -> Result<(), Error> {
+        self.x.assign(region, offset, Some(x.to_le_bytes()))?;
+        self.x_abs
+            .assign(region, offset, Some(x_abs.to_le_bytes()))?;
+        self.is_neg.assign(
+            region,
+            offset,
+            127.into(),
+            u64::from(x.to_le_bytes()[31]).into(),
+        )?;
+        let sum = x.overflowing_add(x_abs).0;
+        self.sum.assign(region, offset, Some(sum.to_le_bytes()))?;
+        self.add_words.assign(region, offset, [x, x_abs], sum)
+    }
+
+    pub(crate) fn x(&self) -> &util::Word<F> {
+        &self.x
+    }
+
+    pub(crate) fn x_abs(&self) -> &util::Word<F> {
+        &self.x_abs
+    }
+
+    pub(crate) fn is_neg(&self) -> &LtGadget<F, 1> {
+        &self.is_neg
+    }
+}
+
+#[cfg(test)]
+mod tests {
+    use super::super::test_util::*;
+    use super::*;
+    use eth_types::Word;
+    use halo2_proofs::halo2curves::bn256::Fr;
+    use halo2_proofs::plonk::Error;
+
+    #[derive(Clone)]
+    struct AbsWordGadgetContainer<F> {
+        absword_gadget: AbsWordGadget<F>,
+    }
+
+    impl<F: Field> MathGadgetContainer<F> for AbsWordGadgetContainer<F> {
+        const NAME: &'static str = "AbsWordGadget";
+
+        fn configure_gadget_container(cb: &mut ConstraintBuilder<F>) -> Self {
+            let absword_gadget = AbsWordGadget::<F>::construct(cb);
+            AbsWordGadgetContainer { absword_gadget }
+        }
+
+        fn assign_gadget_container(
+            &self,
+            input_words: &[Word],
+            region: &mut CachedRegion<'_, '_, F>,
+        ) -> Result<(), Error> {
+            let offset = 0;
+            let x = input_words[0];
+            let x_abs = input_words[1];
+            self.absword_gadget.assign(region, offset, x, x_abs)?;
+
+            Ok(())
+        }
+    }
+
+    #[test]
+    fn test_abs_0_eq_0() {
+        test_math_gadget_container::<Fr, AbsWordGadgetContainer<Fr>>(
+            vec![Word::from(0), Word::from(0)],
+            true,
+        );
+    }
+
+    #[test]
+    fn test_abs_1_eq_1() {
+        test_math_gadget_container::<Fr, AbsWordGadgetContainer<Fr>>(
+            vec![Word::from(1), Word::from(1)],
+            true,
+        );
+    }
+
+    #[test]
+    fn test_abs_1_neq_2() {
+        test_math_gadget_container::<Fr, AbsWordGadgetContainer<Fr>>(
+            vec![Word::from(1), Word::from(2)],
+            false,
+        );
+    }
+
+    #[test]
+    fn test_abs_wordmax_eq_minus1() {
+        test_math_gadget_container::<Fr, AbsWordGadgetContainer<Fr>>(
+            vec![Word::MAX, Word::from(1)],
+            true,
+        );
+    }
```

Comment:
Hmmm some `MAX_HI`/`MAX_LO` testing would be great here as well!
Another thing that I see is that `is_neg(&self)` is actually used externally as well so the test ideally also checks that this flag is set correctly.

Reply:
Added